[SourceKit] Support cancellation of code completion like requests

ahoppen opened the pull request:
This PR performs a few refactorings first (which I could also pull out to a separate PR):
- Moving the cancellation flag from `TypeCheckerOption` to `ASTContext`
- Make `sourcektid_request_handle_t` `const void*`

After that, what remains to be done is just to wire up cancellation tokens and cancellation flags for `CompletionInstance` and make sure to return `CancellableResult::cancelled()` when cancellation is detected.

rdar://83391488

ahoppen committed:
[SourceKit] Make sourcektid_request_handle_t `const void*`

ahoppen commented:
@swift-ci Please smoke test

ahoppen commented:
https://github.com/apple/llvm-project/pull/3548

@swift-ci Please smoke test

ahoppen commented:
apple/llvm-project#3548

@swift-ci Please smoke test

ahoppen commented:
apple/llvm-project#3548

@swift-ci Please smoke test

ahoppen commented:
apple/llvm-project#3548

@swift-ci Please smoke test Windows

rintaro commented on include/swift/Frontend/Frontend.h:
```diff
@@ -107,6 +107,12 @@ class CompilerInvocation {
+  /// the invocation. If the shared pointer is not null and is set to \c true,
+  /// all operations operating on that \c ASTContext are requested to cancel at
+  /// the next possible opportunity.
+  std::shared_ptr<std::atomic<bool>> CancellationFlag = nullptr;
```
What's the motivation of setting this in `CompilerInvocation`? Especially in fast-completion scenario, I feel a little uncomfortable that `CompilerInstance` keep holding the `CancellationFlag` of the _original_ request.

In the fast completion code path, you manually do
```cpp
CI.getASTContext().CancellationFlag = CancellationFlag;
```
I think this is OK, and other ASTManager requests can do this too after constructing ASTContext.

Related: I'm not sure we really need to change `ASTContext::get()`. Other parameters are required because they are references. But `CancellationFlag` can be initialized with `nullptr`.

ahoppen replied:
Two very excellent questions. I think I set `CancellationFlag` on `CompilerInvocation` because that’s what we used to do when it was part of the type checker options. But setting it after creating the compiler instance makes way more sense.

Also: There really is no need to change `ASTContext::get()` I think I wanted to make sure I didn’t miss to set a cancellation flag where I should have set one, but initializing it to `nullptr` and setting it afterwards is much cleaner.

rintaro commented on include/swift/IDE/CompletionInstance.h:
```diff
@@ -129,14 +130,23 @@ class CompletionInstance {
   /// arguments including their order is considered as the same invocation.
   void performOperation(
-      swift::CompilerInvocation &Invocation, llvm::ArrayRef<const char *> Args,
+      swift::CompilerInvocation &&Invocation, llvm::ArrayRef<const char *> Args,
```
I'm not against this change. But is this necessary for this PR? If not, I wouldn't include it in this PR.

ahoppen replied:
No, it’s not. It’s just some cleanup I did while creating the PR. I removed the commit.

ahoppen committed:
[SourceKit] Move the cancellation flag from TypeCheckerOptions to ASTContext

We need to modify the pointer pointing to the cancellation flag when reusing an ASTContext for code completion. This is not possible by the previous design because `TypeCheckerOptions` was `const`. Moving the cancellation flag to `ASTContext` will also allow other stages of the compiler to honor a cancellation request.

ahoppen committed:
[SourceKit] Support cancellation of code completion like requests

Essentially, just wire up cancellation tokens and cancellation flags for `CompletionInstance` and make sure to return `CancellableResult::cancelled()` when cancellation is detected.

rdar://83391488

ahoppen commented:
@swift-ci Please smoke test

ahoppen commented:
Windows failure is unrelated.
